src/java.base/share/classes/com/sun/crypto/provider/HKDFKeyDerivation.java

```diff
@@ -0,0 +1,425 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.crypto.provider;
+
+import javax.crypto.KDFSpi;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.HKDFParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.ByteArrayOutputStream;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import javax.crypto.KDFParameters;
+import java.security.NoSuchAlgorithmException;
+import java.security.ProviderException;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * KDF implementation for the HKDF function.
+ * <p>
+ * This class implements the HKDF-Extract and HKDF-Expand functions from RFC
+ * 5869.  This implementation provides the complete Extract-then-Expand HKDF
+ * function as well as Extract-only and Expand-only variants.
+ */
+abstract class HKDFKeyDerivation extends KDFSpi {
+
+    private final int hmacLen;
+    private final String hmacAlgName;
+
+    private static final int SHA256_HMAC_SIZE = 32;
+    private static final int SHA384_HMAC_SIZE = 48;
+    private static final int SHA512_HMAC_SIZE = 64;
+
+    private static final Integer[] SUPPORTED_HMAC_SIZES = new Integer[]{
+            SHA256_HMAC_SIZE,
+            SHA384_HMAC_SIZE,
+            SHA512_HMAC_SIZE
+    };
+
+    /**
+     * The sole constructor.
+     *
+     * @param kdfParameters
+     *         the initialization parameters (may be {@code null})
+     *
+     * @throws InvalidAlgorithmParameterException
+     *         if the initialization parameters are inappropriate for this
+     *         {@code KDFSpi}
+     */
+    private HKDFKeyDerivation(String hmacAlgName, int hmacLen,
+                              KDFParameters kdfParameters)
+            throws InvalidAlgorithmParameterException {
+        super(kdfParameters);
+        if (kdfParameters != null) {
+            throw new InvalidAlgorithmParameterException(
+                    hmacAlgName + " does not support parameters");
+        }
+        // added to enforce valid values at reviewer's request
+        if (!Arrays.asList(SUPPORTED_HMAC_SIZES).contains(hmacLen)) {
```

Comment:
Since the array of SUPPORTED_HMAC_SIZES is ordered in ascending natural order, you don't need to wrap it as a list to check for contains:
```suggestion
        if (Arrays.binarySearch(SUPPORTED_HMAC_SIZES, hmacLen) < 0) {
```

Reply:
(Also, it could be an `int[]` instead of `Integer[]` in that case)